### ISSUE TITLE
DfsPostOrder traversal on multi rooted graph

### ISSUE THREAD

**Issue by perlindgren** [bluss/petgraph#230]
Hi Folks
I'm new to `petgraph`, so maybe I'm missing the obvious, if so please point me in the right direction.

My goal is to analyse stack usage for programs. To that end, each function is a node, and each edge is a call. 

```
#[derive(Default)]
struct Node {
    name: String,
    local_stack: u32,
    call_stack: u32, // accumulated stack usage
}```

A first approach is to make a traversal of the graph using `DfsPostOrder`, to traverse the graph from the leaves up (mutating the `call_stack` field). However `DfsPostOrder::new` takes a singe root, (while in my case the graph is multi rooted). The algorithm for a single root `ext` is as follows:

```
        let mut dfs = DfsPostOrder::new(&deps, ext);
        while let Some(node) = dfs.next(&deps) {
            // use a detached neighbors walker
            let mut edges = deps.neighbors_directed(node, Incoming).detach();
            while let Some(to) = edges.next_node(&deps) {
                deps[to].call_stack = deps[to]
                    .call_stack
                    .max(deps[to].local_stack + deps[node].call_stack);
            }
        } 
```
One possible way to solve the problem would be to iterate the algorithm for each external:
```
for ext in deps.externals(Incoming) {
    ...
```
but here we get into trouble, `deps` becomes borrowed (and `detach()` seems not implemented). In any case this might be a bad idea, since we start from scratch for each external. The `Dfs` allows to `move_to(&mut self, start: N)`, maybe that is a better approach. 

Maybe there is some other way to operate over the whole graph at once (I saw the map API but got a bit scared).

What would the suggested/idiomatic approach be?
Best regards/
  Per Lindgren
 

Side note 1:
The API provided by petgraph is quite rich, more examples (or even a book) could be very helpful to newcomers like me. 

Side note 2:
Nodes and edges are derived using other tools (Cargo stack-sizes and llvm opt print-callgraph).

**Comment by perlindgren** [comment]
... I implemented the idea with the `move_to`:
```
let mut v = Vec::new();
    for ext in deps.externals(Incoming) {
        v.push(ext);
    }

    let mut dfs = DfsPostOrder::empty(&deps);
    while let Some(ext) = v.pop() {
        dfs.move_to(ext);
        ...
```
It worked out, however I'm still interested in suggested improvements and alternative approaches. (I'm not even sure that I interpreted the API doc correctly, it works as intended for the small tests I made, but it can well be that I missed something...)

Best regards
  Per

**Comment by bluss** [comment]
Hej, that sounds like the best approach as it is now. I can't think of any better way to do it. You have access to the DPO's visit labeling (discovered and finished nodes), so you could implement a manual "scan" to find the remaining externals one by one (keep progress as a node index) every time the current DPO iteration stops, but there's not much difference.

**Comment by perlindgren** [comment]
Thanks for the feedback. 

Maybe the DfsPostOrder::new(&deps, ext)  could take a Vec (or even an iterator) in addition/instead of a single initial node. I figure its quite common that you want to handle multi-rooted graphs.

Best regards/Per

**Comment by bluss** [comment]
I missed one thing, the `stack` field is also public on the DPO so you have all the parts ready. I don't have all of this in my head at the moment, but it looks right to be able to just push all the starts to the stack and then iterate it from there?

In that case you'll just do this:

```
let mut dpo = DfsfPostOrder::empty(&graph);
dpo.stack.extend(graph.externals(..));

// and go
```

Does that work?

**Comment by bluss** [comment]
But yes, why not make a method that takes an iterator that does the equivalent of that. If I'm not wrong about it being correct for all graphs :-)
